File: backend/app/schemas/tenant.py
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict, EmailStr, Field


class TenantProfileResponse(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    industry: str | None = None
    created_at: datetime
    primary_contact_email: str | None = None
    primary_contact_name: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TenantProfileUpdateRequest(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    industry: str | None = None


class TenantSettingsResponse(BaseModel):
    organization_id: uuid.UUID
    timezone: str = "UTC"
    operational_mode: str = "STANDARD"
    default_asset_type: str = "AIRCRAFT"
    notification_email: str | None = None


class TenantSettingsUpdateRequest(BaseModel):
    timezone: str | None = None
    operational_mode: str | None = None
    default_asset_type: str | None = None
    notification_email: EmailStr | None = None


class TenantUserResponse(BaseModel):
    id: uuid.UUID
    email: str
    full_name: str
    is_active: bool
    roles: list[str] = Field(default_factory=list)
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TenantUserRoleUpdateRequest(BaseModel):
    roles: list[str] = Field(..., min_length=1)


class TenantUserStatusUpdateRequest(BaseModel):
    is_active: bool


class TenantRoleInfo(BaseModel):
    role_name: str
    display_name: str
    description: str
    permissions: list[str]
    is_system_role: bool = True


class TenantInvitationRequest(BaseModel):
    email: EmailStr
    full_name: str = Field(..., min_length=1, max_length=255)
    role: str = Field(..., min_length=1)


class TenantInvitationResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    email: str
    full_name: str
    role: str
    status: str  # PENDING, ACCEPTED, EXPIRED, CANCELLED
    created_at: datetime
    expires_at: datetime
    can_resend: bool
    can_cancel: bool


class TenantTeamMember(BaseModel):
    user_id: uuid.UUID
    full_name: str
    email: str
    role: str


class TenantTeamResponse(BaseModel):
    id: str
    name: str
    description: str
    lead_role: str
    member_count: int
    members: list[TenantTeamMember] = Field(default_factory=list)


class TenantUsageMetricItem(BaseModel):
    key: str
    label: str
    tracked: bool
    value: int | None = None
    limit: int | None = None
    is_unlimited: bool = False
    unit: str
    status: str  # TRACKED, NOT_TRACKED, AT_RISK, EXCEEDED


class TenantUsageResponse(BaseModel):
    organization_id: uuid.UUID
    metrics: list[TenantUsageMetricItem]


class TenantDashboardAttentionItem(BaseModel):
    severity: str  # INFO, WARNING, CRITICAL
    category: str
    title: str
    message: str
    link_href: str | None = None
    link_label: str | None = None


class TenantDashboardResponse(BaseModel):
    organization: TenantProfileResponse
    users_count: int
    active_users_count: int
    pending_invitations_count: int
    fleet_count: int
    aircraft_count: int
    drone_count: int
    facility_count: int
    team_count: int
    current_plan: str | None = None
    subscription_status: str | None = None
    effective_features_count: int
    attention_items: list[TenantDashboardAttentionItem] = Field(default_factory=list)

File: frontend/lib/api/tenant.ts
import { apiRequest } from "@/lib/apiClient";
import type { AuditEventResponse } from "@/lib/api/audit";

export interface TenantProfile {
  id: string;
  name: string;
  status: string;
  industry: string | null;
  created_at: string;
  primary_contact_email: string | null;
  primary_contact_name: string | null;
}

export interface TenantProfileUpdatePayload {
  name?: string;
  industry?: string;
}

export interface TenantSettings {
  organization_id: string;
  timezone: string;
  operational_mode: string;
  default_asset_type: string;
  notification_email: string | null;
}

export interface TenantSettingsUpdatePayload {
  timezone?: string;
  operational_mode?: string;
  default_asset_type?: string;
  notification_email?: string;
}

export interface TenantUser {
  id: string;
  email: string;
  full_name: string;
  is_active: boolean;
  roles: string[];
  created_at: string;
}

export interface TenantRoleInfo {
  role_name: string;
  display_name: string;
  description: string;
  permissions: string[];
  is_system_role: boolean;
}

export interface TenantInvitation {
  id: string;
  user_id: string;
  email: string;
  full_name: string;
  role: string;
  status: "PENDING" | "ACCEPTED" | "EXPIRED" | "CANCELLED";
  created_at: string;
  expires_at: string;
  can_resend: boolean;
  can_cancel: boolean;
}

export interface TenantInvitationPayload {
  email: string;
  full_name: string;
  role: string;
}

export interface TenantTeamMember {
  user_id: string;
  full_name: string;
  email: string;
  role: string;
}

export interface TenantTeam {
  id: string;
  name: string;
  description: string;
  lead_role: string;
  member_count: number;
  members: TenantTeamMember[];
}

export interface TenantUsageMetric {
  key: string;
  label: string;
  tracked: boolean;
  value: number | null;
  limit: number | null;
  is_unlimited: boolean;
  unit: string;
  status: "TRACKED" | "NOT_TRACKED" | "AT_RISK" | "EXCEEDED";
}

export interface TenantUsage {
  organization_id: string;
  metrics: TenantUsageMetric[];
}

export interface TenantDashboardAttentionItem {
  severity: "INFO" | "WARNING" | "CRITICAL";
  category: string;
  title: string;
  message: string;
  link_href?: string;
  link_label?: string;
}

export interface TenantDashboardStats {
  organization: TenantProfile;
  users_count: number;
  active_users_count: number;
  pending_invitations_count: number;
  fleet_count: number;
  aircraft_count: number;
  drone_count: number;
  facility_count: number;
  team_count: number;
  current_plan: string | null;
  subscription_status: string | null;
  effective_features_count: number;
  attention_items: TenantDashboardAttentionItem[];
}

export const tenantApi = {
  getDashboard: (accessToken: string) =>
    apiRequest<TenantDashboardStats>("/tenant/dashboard", { accessToken }),

  getProfile: (accessToken: string) =>
    apiRequest<TenantProfile>("/tenant/profile", { accessToken }),

  updateProfile: (accessToken: string, payload: TenantProfileUpdatePayload) =>
    apiRequest<TenantProfile>("/tenant/profile", {
      accessToken,
      method: "PATCH",
      body: payload,
    }),

  getSettings: (accessToken: string) =>
    apiRequest<TenantSettings>("/tenant/settings", { accessToken }),

  updateSettings: (accessToken: string, payload: TenantSettingsUpdatePayload) =>
    apiRequest<TenantSettings>("/tenant/settings", {
      accessToken,
      method: "PATCH",
      body: payload,
    }),

  // Platform Control Plane: soft-deletes this admin's OWN organization
  // (backend/app/services/deletion_service.py's request_organization_
  // deletion). There is no tenant-facing undo -- every subsequent request,
  // including this admin's own current session, is refused immediately
  // afterward. Only Platform Admin can restore or approve permanent
  // deletion (lib/api/deletion.ts).
  requestDeletion: (accessToken: string, reason?: string) =>
    apiRequest<{ message: string }>("/tenant/deletion-request", {
      accessToken,
      method: "POST",
      body: { reason },
    }),

  listUsers: (accessToken: string) =>
    apiRequest<TenantUser[]>("/tenant/users", { accessToken }),

  getUser: (accessToken: string, userId: string) =>
    apiRequest<TenantUser>(`/tenant/users/${userId}`, { accessToken }),

  updateUserRoles: (accessToken: string, userId: string, roles: string[]) =>
    apiRequest<TenantUser>(`/tenant/users/${userId}/roles`, {
      accessToken,
      method: "PATCH",
      body: { roles },
    }),

  updateUserStatus: (accessToken: string, userId: string, isActive: boolean) =>
    apiRequest<TenantUser>(`/tenant/users/${userId}/status`, {
      accessToken,
      method: "PATCH",
      body: { is_active: isActive },
    }),

  listRoles: (accessToken: string) =>
    apiRequest<TenantRoleInfo[]>("/tenant/roles", { accessToken }),

  listInvitations: (accessToken: string) =>
    apiRequest<TenantInvitation[]>("/tenant/invitations", { accessToken }),

  inviteUser: (accessToken: string, payload: TenantInvitationPayload) =>
    apiRequest<TenantInvitation>("/tenant/invitations", {
      accessToken,
      method: "POST",
      body: payload,
    }),

  resendInvitation: (accessToken: string, userId: string) =>
    apiRequest<{ message: string }>(`/tenant/invitations/${userId}/resend`, {
      accessToken,
      method: "POST",
    }),

  cancelInvitation: (accessToken: string, userId: string) =>
    apiRequest<{ message: string }>(`/tenant/invitations/${userId}/cancel`, {
      accessToken,
      method: "POST",
    }),

  listTeams: (accessToken: string) =>
    apiRequest<TenantTeam[]>("/tenant/teams", { accessToken }),

  getUsage: (accessToken: string) =>
    apiRequest<TenantUsage>("/tenant/usage", { accessToken }),

  listAuditEvents: (
    accessToken: string,
    params: { limit?: number; offset?: number } = {}
  ) => {
    const qs = new URLSearchParams();
    if (params.limit !== undefined) qs.set("limit", String(params.limit));
    if (params.offset !== undefined) qs.set("offset", String(params.offset));
    const query = qs.toString();
    return apiRequest<AuditEventResponse[]>(
      `/tenant/audit${query ? `?${query}` : ""}`,
      { accessToken }
    );
  },
};
